app/code/Magento/Eav/Model/Entity/Attribute.php

```diff
@@ -253,6 +253,16 @@ public function beforeSave()
             );
         }
 
+        /**
+         * Check attribute_code for allowed characters
+         */
+        if (isset($this->_data['attribute_code'])
+            && !preg_match('/^[a-z]+[a-z0-9_]+$/', $this->_data['attribute_code'])) {
```

Comment:
Model is not right place for validation, there is some validation already present in controller, isn't it?

Reply:
@orlangur I suggest to use this validation also when we need to create some eav attributes programmatically. As for the controllers, there is a frontend validation in the backend while an attribute is created.

Reply:
@eduard13 frontend validation is not enough. I'm pretty sure this is validated on PHP side also. Validation should not be added to model itself, when you create attribute programmatically - you should validate attribute code prior to passing it into model.

Reply:
@orlangur I agree with you that we should check the attribute code before saving it, but sometimes you don't know that some chars aren't allowed, like . (dots), and you successfully save the attribute, but later you aren't able to properly work with the attribute (because the getters and setters aren't well generated).

By the way, if we'll take a look closer in this method, there are other several validation rules, so it is needed to refactor this method and move all the validations to a separate file?

Reply:
@eduard13 didn't check the code out of your changeset :) I need some time to understand the whole picture, please be patient.

Reply:
Hi @orlangur any suggestions here?